src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,555 @@
+        string calldata socket
+    ) external onlyWhenNotPaused(PAUSED_REGISTER_OPERATOR) {
+        _registerOperatorWithCoordinatorAndNoOverfilledQuorums(msg.sender, quorumNumbers, pubkey, socket);
+    }
+
+    /**
+     * @notice Registers msg.sender as an operator with the middleware when the quorum operator limit is full. To register 
+     * while maintaining the limit, the operator chooses another registered opeerator with lower stake to kick.
+     * @param quorumNumbers are the bytes representing the quorum numbers that the operator is registering for
+     * @param pubkey is the BLS public key of the operator
+     * @param operatorKickParams are the parameters for the deregistration of the operator that is being kicked from each 
+     * quorum that will be filled after the operator registers. These parameters should include an operator, their pubkey, 
+     * and ids of the operators to swap with the kicked operator. 
+     * @param signatureWithSaltAndExpiry is the signature of the churnApprover on the operator kick params with a salt and expiry
+     */
+    function registerOperatorWithCoordinator(
+        bytes calldata quorumNumbers, 
+        BN254.G1Point memory pubkey,
+        string calldata socket,
+        OperatorKickParam[] calldata operatorKickParams,
+        SignatureWithSaltAndExpiry memory signatureWithSaltAndExpiry
+    ) external onlyWhenNotPaused(PAUSED_REGISTER_OPERATOR) {
+        // register the operator
+        uint32[] memory numOperatorsPerQuorum = _registerOperatorWithCoordinator(msg.sender, quorumNumbers, pubkey, socket);
+
+        // get the registering operator's operatorId and set the operatorIdsToSwap to it because the registering operator is the one with the greatest index
+        bytes32[] memory operatorIdsToSwap = new bytes32[](1);
+        operatorIdsToSwap[0] = pubkey.hashG1Point();
+
+        // verify the churnApprover's signature
+        _verifyChurnApproverSignatureOnOperatorChurnApproval(operatorIdsToSwap[0], operatorKickParams, signatureWithSaltAndExpiry);
+
+        // kick the operators
+        for (uint256 i = 0; i < quorumNumbers.length; i++) {
+            // check that the quorum has reached the max operator count
+            uint8 quorumNumber = uint8(quorumNumbers[i]);
+            OperatorSetParam memory operatorSetParam = _quorumOperatorSetParams[quorumNumber];
+            {
+                // if the number of operators for the quorum is less than or equal to the max operator count, 
+                // then the quorum has not reached the max operator count
+                if(numOperatorsPerQuorum[i] <= operatorSetParam.maxOperatorCount) {
+                    continue;
+                }
+
+                require(
+                    operatorKickParams[i].quorumNumber == quorumNumber, 
+                    "BLSRegistryCoordinatorWithIndices.registerOperatorWithCoordinator: quorumNumber not the same as signed"
+                );
```

Comment:
am I reading this right that the `operatorKickParams` array needs to be the same length as the number of quorums the operator is registering for?
seems like this check would fail, otherwise.
if this is a requirement, we should probably just have a simple length check higher up.
if we don't want it to be a requirement, then we need something like a second index here, but I think the logic could get messy.

**possible security concern:** if these have to be the same length, then the `churnApprover` may need to sign arrays containing some empty entries -- I'm concerned that we handle these correctly such signed arrays under all circumstances.

Reply:
to elaborate on the potential security concern, with an example:

- operator wants to register for quorums [1,3] and currently only quorum 3 is full
- churnApprover signs array with [empty other than quorum number, empty other than quorum number, full details of operatorToKick] entries
- new operator registers for quorum 1, making it full
- original operator calls this function, and we end up hitting a case where we try to kick an operator based on the signed entry that's empty other than quorum number

I guess this might be more of a race condition, actually? If we are enforcing behavior well on deregistrations then we should at least just hit a revert in this scenario

Reply:
will, e.g., the call to `getCurrentOperatorStakeForQuorum` below revert? if not then I think there could be an issue here

Reply:
> the arrays don't need to be the same length? it just checks that the entry in the array matches with the quorum that has met the cap?

But it checks the `i`-th entry in the array of `operatorKickParams`. So like, if I'm registering for 12 quorums and need to kick someone only in the 12th quorum, I need to submit `operatorKickParams` of length 12, because this code will look at the 12th entry in the array.
I think this is "OK", just perhaps a tad counterintuitive(?)

Reply:
Great find again. Here: https://github.com/Layr-Labs/eigenlayer-contracts/pull/35/commits/b9549ac23785fde8d3abca15702f2ee73fbfddd7

Reply:
LGTM, thanks.